Player transfer bug fixes

garrettluskey opened the pull request:
Many issues with hero transfer fixed in this. Object Transaction system was scrapped for this update and will remain as an on-hold until we see more value in it. SyncStore works fine for now. Mainly serialization updates and manager updates to pass around newly created heros.

garrettluskey committed:
Character is transferable via save xfer

garrettluskey committed:
Merge branch 'development' into Player-transfer-bug-fixes

garrettluskey committed:
Updated serializer to assign Id and client manager starts new player in training field

garrettluskey committed:
Changed for loop to parallel foreach and removed debug if statements

garrettluskey committed:
Merge branch 'development' into Player-transfer-bug-fixes

araex commented on source/Coop/Mod/Managers/ClientManager.cs:
```diff
@@ -1,51 +1,41 @@
         {
             base.OnLoadFinished();
-
+            Parallel.ForEach(MobileParty.All, (party) =>
```
Feels kinda weird to express it that way. The code is not actually thread save, it only works under the assumption that only 1 party.Name == m_PartyName.

How about this instead:
```C#
MobileParty playerParty = MobileParty.All.AsParallel().SingleOrDefault(party => party.Name.ToString() == m_PartyName);
if(playerParty != null)
{
   // ...
}
```

garrettluskey replied:
I agree

araex commented on source/Coop/Mod/Serializers/CustomSerializers/PlayerHeroSerializer.cs:
```diff
@@ -159,7 +159,25 @@ public override object Deserialize()
+
+            HeroDeveloper newDeveloper = (HeroDeveloper)ctor.Invoke(new object[] { hero });
+            hero.GetType()
+                .GetField("_heroDeveloper", BindingFlags.NonPublic | BindingFlags.Instance)
```
Do you happen to know of a way to get this to yield a compile error if the field does not exist? Maybe trough a public property that uses the field internally?

Since they are refactoring game DLLs, these hard coded variable names may compile, but no longer work in a new game version. Some way to at least notice that at compile time would be great.

garrettluskey replied:
.SetValue will throw a compiler exception due to FieldInfo being null

araex reviewed (approved):
Looks good

araex committed:
- newest update removed TaleWorlds.CampaignSystem.Actions.RecruitAction.
The patch is obsolete
- Fixed an include issue, likely due to the newest update

araex commented:
I pushed 2 compile fixes in 32ffda5 . Most likely these are due to the latest update.

garrettluskey committed:
Updated to handle null from search

garrettluskey committed:
Merge branch 'Player-transfer-bug-fixes' of https://github.com/Bannerlord-Coop-Team/BannerlordCoop into Player-transfer-bug-fixes
